goalie/function_data.py

```diff
@@ -30,52 +30,121 @@ def __init__(self, time_partition, function_spaces):
             }
         )
 
     @property
-    def data(self):
+    def _data_by_field(self):
+        """
+        Extract field data array in the default layout: as a doubly-nested dictionary
+        whose first key is the field name and second key is the field label. Entries
+        of the doubly-nested dictionary are doubly-nested lists, indexed first by
+        subinterval and then by export.
+        """
         if self._data is None:
             self._create_data()
         return self._data
 
     def __getitem__(self, key):
-        return self.data[key]
+        return self._data_by_field[key]
 
     def items(self):
-        return self.data.items()
-
+        return self._data_by_field.items()
 
-class SolutionData(FunctionData, abc.ABC):
-    """
-    Abstract base class that defines the API for solution data classes.
-    """
+    @property
+    def _data_by_label(self):
+        """
+        Extract field data array in an alternative layout: as a doubly-nested dictionary
+        whose first key is the field label and second key is the field name. Entries
+        of the doubly-nested dictionary are doubly-nested lists, which retain the default
+        layout: indexed first by subinterval and then by export.
+        """
+        tp = self.time_partition
+        return AttrDict(
+            {
+                label: AttrDict(
+                    {field: self._data_by_field[field][label] for field in tp.fields}
+                )
+                for label in self.labels
+            }
+        )
 
     @property
-    def solutions(self):
-        return self.data
+    def _data_by_subinterval(self):
+        """
+        Extract field data array in an alternative format: as a list indexed by
+        subinterval. Entries of the list are doubly-nested dictionaries, which retain
+        the default layout: with the first key being field label and the second key being
+        the field name. Entries of the doubly-nested dictionaries are lists of field
+        data, indexed by export.
+        """
+        tp = self.time_partition
+        return [
+            AttrDict(
+                {
+                    field: AttrDict(
+                        {
+                            label: self._data_by_field[field][label][subinterval]
+                            for label in self.labels
+                        }
+                    )
+                    for field in tp.fields
+                }
+            )
+            for subinterval in range(tp.num_subintervals)
+        ]
+
+    def extract(self, layout="field"):
+        """
+        Extract field data array in a specified layout.
+
+        The default layout is a doubly-nested dictionary whose first key is the field
+        name and second key is the field label. Entries of the doubly-nested dictionary
+        are doubly-nested lists, indexed first by subinterval and then by export. That
+        is: ``data[field][label][subinterval][export]``.
```

Comment:
Would something like "export_index" maybe be better than "export"? I feel like "export" is not very informative. Feel free to ignore if you disagree :)

Reply:
Personally, I prefer "export" for conciseness. It makes as much sense as saying "subinterval" rather than "subinterval_index".